### PR TITLE
Iter13

### PR DESCRIPTION
**Pull request by evgen3000**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+make create-db:
+	docker run --name admin -p "5432:5432" -e POSTGRES_DB=admin -e POSTGRES_USER=admin -e POSTGRES_PASSWORD=admin -d "postgres:14-bullseye"
+	until docker exec admin pg_isready -U admin
```
Comment (arttet) [inline Makefile:3]: [CHANGE REQUEST] I am unsure that it needs.

Comment (arttet) [reply Makefile:3]: [SUGGESTION] It would be better if you used `docker compose`.

File: cmd/server/handlers/handlers.go [hunk 2]
```diff
@@ -6,22 +6,25 @@ import (
 	"strconv"
 	"strings"
 
-	"evgen3000/go-musthave-metrics-tpl.git/cmd/server/storage/memstorage"
+	"evgen3000/go-musthave-metrics-tpl.git/cmd/server/storage"
```
Comment (arttet) [inline cmd/server/handlers/handlers.go:11]: [CHANGE REQUEST] Please improve your layout.
https://github.com/golang-standards/project-layout

File: cmd/server/postgres/postgres.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	if err != nil {
+		log.Fatalf("Unable to parse config: %v", err)
+	}
+
+	pool, poolErr := pgxpool.NewWithConfig(context.Background(), config)
+	if poolErr != nil {
+		log.Fatalf("Unable to connect to database: %v", err)
+	}
+
+	q := `CREATE TABLE gauge
```
Comment (arttet) [inline cmd/server/postgres/postgres.go]: [CHANGE REQUEST] Please use migrations.

File: cmd/server/storage/dbstorage/dbstorage.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+	ID    string `json:"id"`
+	Value *int64 `json:"value"`
+}
+
+func (db *DBStorage) StorageType() string {
+	return "db"
+}
+
+func (db *DBStorage) SetMetrics(metrics []dto.MetricsDTO) {
+	tx, err := db.Pool.Begin(context.Background())
```
Comment (arttet) [inline cmd/server/storage/dbstorage/dbstorage.go:50]: [CHANGE REQUEST] `Context` should be a parameter of this function.

File: cmd/server/storage/dbstorage/dbstorage.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+		if err != nil {
+			if rollbackErr := tx.Rollback(context.Background()); rollbackErr != nil {
+				log.Fatalf("Unable to rollback transaction: %v", rollbackErr)
+			}
+		}
+	}()
+
+	for _, metric := range metrics {
+		if metric.MType == dto.MetricTypeGauge && metric.Value != nil {
+			q := `INSERT INTO public.gauge (id, value)
```
Comment (arttet) [inline cmd/server/storage/dbstorage/dbstorage.go]: [CHANGE REQUEST] Please separate Go and SQL code.

File: cmd/server/storage/storage.go [hunk 6]
```diff
@@ -1,30 +1,51 @@
...
-	"evgen3000/go-musthave-metrics-tpl.git/cmd/server/storage/filemanager"
+	"evgen3000/go-musthave-metrics-tpl.git/cmd/server/postgres"
+	"evgen3000/go-musthave-metrics-tpl.git/cmd/server/storage/dbstorage"
 	"evgen3000/go-musthave-metrics-tpl.git/cmd/server/storage/memstorage"
+	"evgen3000/go-musthave-metrics-tpl.git/cmd/server/storage/memstorage/filemanager"
+	"evgen3000/go-musthave-metrics-tpl.git/internal/dto"
 )
 
-type MemStorageConfig struct {
+type Interface interface {
```
Comment (arttet) [inline cmd/server/storage/storage.go]: [CHANGE REQUEST] Please rename it.

**Review by arttet — CHANGES_REQUESTED** [review]
Please add more tests

File: Makefile (hunk 1 above)
Comment (evgen3000) [reply Makefile:3]: I use it only for local developing. It useful for fast run a code

File: cmd/server/handlers/handlers.go (hunk 2 above)
Comment (evgen3000) [reply cmd/server/handlers/handlers.go:11]: What do you mean? I guess, this layout really simple and understandable. Could you tell me what is wrong?

Comment (arttet) [reply cmd/server/handlers/handlers.go:11]: The handlers package should be in the `internal` package. The `cmd` package should be simple.

File: cmd/server/handlers/handlers.go [hunk 7]
```diff
@@ -1,38 +1,45 @@
...
+	} else {
+		rw.WriteHeader(http.StatusInternalServerError)
+	}
 }
 
 func (h *Handler) HomeHandler(rw http.ResponseWriter, _ *http.Request) {
 	var body strings.Builder
 	body.WriteString("<h4>Gauges</h4>")
-	for gaugeName, value := range h.Storage.GetAllGauges() {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```
Comment (arttet) [inline cmd/server/handlers/handlers.go:35]: [CHANGE REQUEST] It should be in the `middleware` package.

File: cmd/server/storage/dbstorage/dbstorage.go [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+		if err != nil {
+			if rollbackErr := tx.Rollback(context.Background()); rollbackErr != nil {
+				log.Fatalf("Unable to rollback transaction: %v", rollbackErr)
+			}
+		}
+	}()
+
+	for _, metric := range metrics {
+		if metric.MType == dto.MetricTypeGauge && metric.Value != nil {
+			err = db.InsertOrUpdateGauge(ctx, metric.ID, *metric.Value)
```
Comment (arttet) [inline cmd/server/storage/dbstorage/dbstorage.go:65]: [SUGGESTION] You can optimize these queries.

Comment (arttet) [inline cmd/server/storage/dbstorage/dbstorage.go:65]: This is `upsert`.

File: cmd/server/storage/dbstorage/dbstorage.go [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+	err = tx.Commit(context.Background())
+	if err != nil {
+		log.Fatalf("Unable to commit transaction: %v", err)
+	}
+}
+
+func (db *DBStorage) SetGauge(ctx context.Context, metricName string, value float64) {
+	var gauge Gauge
+	q := `SELECT id, value FROM public.gauge WHERE id = $1;`
```
Comment (arttet) [inline cmd/server/storage/dbstorage/dbstorage.go:87]: [SUGGESTION] You can do it in one query.

**Review by arttet — APPROVED** [review]

**Pull request merged**